### ISSUE TITLE
Why can I create data objects with mismatch in in x and y shape?

### ISSUE THREAD

**Issue by hamogu** [sherpa/sherpa#895]
All of the below commands succeed in my current branch (master with the PRs on new XSPEC version pulled in which don't touch the Sherpa data inititialization).
I had expected all of them to fail because the length of x and y arrays don't match.

```
import numpy as np
from sherpa.data import Data1DInt

list4 = [1,2,3,4]
arr4 = np.arange(4)
list5 = [1,2,3,4,5]
arr5 = np.arange(5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=arr5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=list5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=arr5, staterror=arr5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=list5, staterror=list5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=list4, staterror=list5)

d = Data1DInt(name='test', xlo=arr4, xhi=arr4 + 1,
              y=list5, staterror=list4)
```

Maybe the logic that tests matching shapes in in the `load_array1 function on the UI level, but I consider it a bug that I can do that on the OO level. If that's the design intent, then someone needs to remove the "bug" label.
It cost me about 30 min of debugging time on a science project, because I got a downstream error when fitting such mismatched data object and it took me a long time to figure out that sherpa would allow me to create such an object in the first place.

**Comment by DougBurke** [comment]
My guess (from having looked at the data hierarchy in the past, not for this particular issue) is that the validation is in the wrong place, and should be in the Data class and not the UI layer. Historically we weren't exposing the lower layers, and just concentrating on the UI layer, but I believe that we should be pushing checks down as much as we can.

**Comment by DougBurke** [comment]
A comment for when we come to look at this: should we allow the DataXXX classes the ability to change the independent axes of the dataset (e.g. to add or remove bins) or should we make the independent axes read-only? At the moment we let users change the independent axes - i.e. there's a `@indep.setter` decorator on `sherpa.data.Data.indep` - but I wonder if we would lose much functionality if we make `indep` a read-only attribute?

The reason for asking is that if the independent axes are fixed when the object is created then we can simplify checks for array sizes (that is, the DataXXX object knows the size and can enforce it when any related value, such as `syserror`, is set). If we allow the independent axes to change then we have to revalidate the whole object when the axes are changed, and then get into a problem of "I need to change the syserror array to match the new data, but this is only possible after changing the value, at which point the syserror array is now invalid"; or perhaps changing the independent axes just sets all related values to None? I think making the independent axis as read-only greatly simplifies things, but may have some unintended consequences.

**Comment by hamogu** [comment]
I like the idea of making the indep axes read-only, I think the simplicity is worth the extra work of creating a new object in the rare cases that someone actually needs to change it (see #887 for en example that allows to do something similar, but is never used). In practice that probably means we still have a `@indep.getter`, but then keep the actual data in `_indep` (or similar). If there is ever reason to screw around with the axis (e.g. during testing and debugging), you can do it, but you are clearly on your own.

**Comment by hamogu** [comment]
I also note that this is the type of design change that needs to be documented (@wmclaugh !), so if we do something like that, I suggest to add a docs page "Why we chose to make the indep axis a read-only attribute" so have that reasoning explained for future developers, but also for users.

**Comment by DougBurke** [comment]
For fun I ran the test suite after removing set_indep and @indep.setter and got a number of failures. I haven't dug through them (I expect some of them to be functional changes) but there's an interesting one: the set_psf call fails below, which implies any restriction would need to rewrite some code (e.g. creating new DataXXX objects rather than over-writing them, which may be tricky depending on how it's used). 

So definitely not a quick change!

```
    def test_calc_source_model_sum2d(setup):
        ui, image, psf_bin05, _ = setup

        ui.load_image(image)
        ui.get_data().sky.cdelt = (1, 1)

        ui.load_psf('psf', psf_bin05)
>       ui.set_psf('psf')
```

**Comment by DougBurke** [comment]
For thinking about, but not right now: @hamogu and I have recently-ish discussed whether `DataPHA("blah", None, None)` should be allowed or not. If we are always validating things then we need to allow the case of "one or other of the independent and dependent axes are not set", which is feasible, but it means that if you want to change the number of data points in a data object you would need to set one of the two to `None` first, otherwise you can't change things. We could have something that allows you to set both at the same time, but in reality you'd also need to be able to change staterror, syserror, and the myriad PHA-related attributes. We could imagine some sort of a "reset" method which clears out (i.e. sets to None) these attributes without changing other fields/metadata, but use cases are needed.

**Comment by hamogu** [comment]
For the specific case of this issue, we can probably get 95 % there with 5% of the work. How about we just validate in `__init__` like this (pseudo-code):
```
if x is not None and y is not None and (len(x) != len(y)):
    raise Error("x and y need to have same length")
```
As a user, I've never knowingly changed the x or y values of any data object - I just create a new one. The check above would not stop the sherpa code itself from making changes (like the PSF case that @DougBurke discovered above), nor would it stop a really determined user, but it would protect the casual user (like me) who just accidentally passes in arrays of the wrong size.

**Comment by hamogu** [comment]
I think this has been addressed in #1477
